nextest-runner/src/test_output.rs

```diff
@@ -0,0 +1,322 @@
+    #[inline]
+    fn poll_fill_buf(
+        self: Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<std::io::Result<&[u8]>> {
+        match self.project() {
+            StreamProj::Stdout { o } => o.poll_fill_buf(cx),
+            StreamProj::Stderr { o } => o.poll_fill_buf(cx),
+        }
+    }
+
+    #[inline]
+    fn consume(self: Pin<&mut Self>, amt: usize) {
+        match self.project() {
+            StreamProj::Stdout { o } => o.consume(amt),
+            StreamProj::Stderr { o } => o.consume(amt),
+        }
+    }
+}
+
+/// The size of each buffered reader's buffer, and the size at which we grow
+/// the interleaved buffer.
+///
+/// This size is not totally arbitrary, but rather the (normal) page size on
+/// most linux, windows, and macos systems.
+const CHUNK_SIZE: usize = 4 * 1024;
+
+/// Collects the stdout and/or stderr streams into a single buffer
+#[allow(clippy::needless_lifetimes, clippy::let_and_return)]
+pub fn collect_test_output<'a>(
+    stdout: Option<tokio::process::ChildStdout>,
+    stderr: Option<tokio::process::ChildStderr>,
+    acc: &'a mut TestOutputAccumulator,
+) -> impl futures::Future<Output = Result<(), crate::errors::CollectTestOutputError>> + 'a {
+    let read_loop = async move {
+        // Currently, the runner either captures both, or neither, never just one
+        let (mut stdout, mut stderr) = match (stdout, stderr) {
+            (Some(out), Some(err)) => (
+                Stream::Stdout {
+                    o: BufReader::with_capacity(CHUNK_SIZE, out),
+                },
+                Stream::Stderr {
+                    o: BufReader::with_capacity(CHUNK_SIZE, err),
+                },
+            ),
+            (None, None) => {
+                return Ok(());
+            }
+            (out, err) => {
+                unreachable!(
+                    "one of stdout: {out:?} stderr: {err:?}, was captured, but not the other"
+                );
+            }
+        };
+
+        let mut out_done = false;
+        let mut err_done = false;
+
+        while !out_done || !err_done {
+            let (read, which) = tokio::select! {
+                buf = stdout.fill_buf() => {
+                    let buf = buf.map_err(crate::errors::CollectTestOutputError::ReadStdout)?;
+                    push_chunk(acc, buf, true);
+                    (buf.len(), true)
+                }
+                buf = stderr.fill_buf() => {
+                    let buf = buf.map_err(crate::errors::CollectTestOutputError::ReadStderr)?;
+                    push_chunk(acc, buf, false);
+                    (buf.len(), false)
+                }
+            };
+
+            if which {
+                stdout.consume(read);
+                out_done = read == 0;
+            } else {
+                stderr.consume(read);
+                err_done = read == 0;
+            }
+        }
+
+        Ok(())
+    };
+
+    read_loop
```

Comment:
You can return that future directly without binding it to a variable.

Reply:
Not without lifetime errors.

Reply:
That's...strange because the return type has the right lifetime annotation.